### PR TITLE
Added toggle button widget

### PR DESCRIPTION
**Pull request by vinothvino42**
Ticket: #209 
In this PR, I added a single selection toggle button widget

Sample YAML:
```
View:
  styles: { useSafeArea: true }
  body:
    Column:
      styles: { gap: 50, useSafeArea: true }
      children:
        - ToggleButton1
        - ToggleButton2
ToggleButton1:
  body:
    Column:
      children:
        - Text:
            text: "Toggle Button 1"
        - ToggleButtons:
            styles:
              {
                gap: 20,
                color: blue,
                selectedColor: pink,
                selectedBackgroundColor: green,
              }
            value: Blue
            items:
              - Red
              - Blue
              - Green
            onChange: |
              //@code
              onChangeDemo.text = 'You selected 1: ' +this.value;

        - Text:
            styles: { padding: 8, fontSize: 18 }
            text: Select Value 1
            id: onChangeDemo

ToggleButton2:
  body:
    Column:
      children:
        - Text:
            text: "Toggle Button 2"
        - ToggleButtons:
            styles:
              {
                gap: 8,
                color: blue,
                selectedColor: pink,
                selectedBackgroundColor: green,
              }
            value: green
            items:
              - {
                  "label": "Red",
                  "value": "red",
                  "icon": { "name": "home", "size": 24 },
                }
              - { "label": "Blue", "value": "blue", "icon": { "name": "star" } }
              - { "label": "Green", "value": "green" }
            onChange: |
              //@code
              onChangeDemo.text = 'You selected 2: ' +this.value;

        - Text:
            styles: { padding: 8, fontSize: 18 }
            text: Select Value 2
            id: onChangeDemo

```

### CODE REVIEW

**Comment by vusters** [comment]
Please also add this to ensemble-schema. This is how users will know how to use the widget.

**Review by vusters — APPROVED** [review]

File: lib/widget/toggle_button.dart [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+  bool isIconExist(List v) {
+    for (var e in v) {
+      if (e['icon'] != null) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void calculateIconSize(List v) {
```
[the remark this answers is not in the document]
Comment (vinothvino42) [reply lib/widget/toggle_button.dart:119]: Okay I'll remove it

File: lib/widget/toggle_button.dart [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+class EnsembleToggleButtonState extends WidgetState<EnsembleToggleButton> {
+  List<ToggleItem>? _items = [];
+
+  @override
+  void initState() {
+    super.initState();
+    _setValue();
+  }
+
+  void _setValue() {
```
Comment (evshi) [inline lib/widget/toggle_button.dart:153]: Why do we need both `_setValue` and `_updatedSelectedState`?

File: lib/widget/toggle_button.dart [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+      final valueIndex = widget.controller.items
+          ?.indexWhere((element) => element.value == value);
+      widget.controller.items!.asMap().forEach((index, value) {
+        _items!.add(
+          ToggleItem(
+            label: value.label,
+            value: value.value,
+            icon: value.icon,
+            isIcon: value.isIcon,
+            isSelected: (valueIndex != null && valueIndex != -1)
```
Comment (evshi) [inline lib/widget/toggle_button.dart:165]: nit: `index` is never null or -1 so you don't need this check

File: lib/widget/toggle_button.dart [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+      fillColor: controller.selectedBackgroundColor,
+      borderColor: controller.borderColor,
+      selectedBorderColor: controller.selectedBorderColor,
+      constraints: const BoxConstraints(
+        minHeight: 40.0,
+        minWidth: 80.0,
+      ),
+      children: List.generate(
+        _items!.length,
+        (index) => _buildWidget(_items![index]),
```
Comment (evshi) [inline lib/widget/toggle_button.dart:193]: Any reason to use LIst.generate instead of map?

File: lib/widget/toggle_button.dart [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+      ),
+      isSelected: [
+        ..._items!.map((e) => e.isSelected).toList(),
+      ],
```
Comment (evshi) [inline lib/widget/toggle_button.dart:197]: nit: extra list being created here?

File: assets/schema/ensemble_schema.json [hunk 6]
```diff
@@ -4,24 +4,27 @@
...
+    "Import": {},
     "ViewGroup": {
       "$ref": "#/$defs/Menu",
       "description": "Group multiple Views together and put them behind a menu."
     },
     "View": {
       "type": "object",
       "description": "This is your root View. It requires a body widget.",
-      "required": ["body"],
+      "required": [
```
Comment (evshi) [inline assets/schema/ensemble_schema.json:15]: Do these formatting changes have any consequences @vusters ?

File: lib/widget/toggle_button.dart (hunk 4 above)
Comment (vinothvino42) [reply lib/widget/toggle_button.dart:193]: You'll get an index when iterating the list with `List.generate`

File: lib/widget/toggle_button.dart (hunk 5 above)
Comment (vinothvino42) [reply lib/widget/toggle_button.dart:197]: Yeah. For the isSelected, we need to pass an array of booleans for the selected/unselected state of an item

File: lib/widget/toggle_button.dart (hunk 3 above)
Comment (vinothvino42) [reply lib/widget/toggle_button.dart:165]: I'm checking the value is present in the items (list of type SelectOneItem) and returning its index using the `indexWhere` method. It returns -1 if there are no items found and the controller items are nullable lists, so you'll get an optional value here. It's better to handle it safely

File: lib/widget/toggle_button.dart (hunk 2 above)
Comment (vinothvino42) [reply lib/widget/toggle_button.dart:153]: `_setValue` is for creating an initial items list and its selection based on the value. `_updatedSelectedState` is used to update the existing list and its selection state based on the index value

File: lib/widget/toggle_button.dart (hunk 4 above)
Comment (snehmehta) [reply lib/widget/toggle_button.dart:172]: fyi: inside core collection package, we have `mapIndexed` you'll get both index and element.

Comment (vinothvino42) [reply lib/widget/toggle_button.dart:172]: > fyi: inside core collection package, we have `mapIndexed` you'll get both index and element.

Do you think in this case `mapIndexed` is useful?

**Comment by vinothvino42** [comment]
@vusters @evshi I just changed the default toggle button to a custom toggle button. Can you please take a look at it?

Sample YAML with four different styles of ToggleButtons

<details>
  <summary>Click me</summary>

  ### YAML Code
  ```yaml
  View:
  styles: { useSafeArea: true, scrollableView: true }
  body:
    Column:
      styles: { gap: 30, padding: 10 }
      crossAxis: stretch
      children:
        - ToggleButton1
        - ToggleButton2
        - ToggleButton3
        - ToggleButton4
ToggleButton1:
  body:
    Column:
      styles: { gap: 10 }
      crossAxis: center
      children:
        - Text:
            styles: { fontSize: 18 }
            text: "Toggle Button 1 - Without Styles"
        - ToggleButton:
            value: Blue
            items:
              - Red
              - Blue
              - Green
              - Orange
              - Purple
              - Yellow
            onChange: |
              //@code
              onChangeDemo.text = 'You selected toggle button 1 value: ' +this.value;

        - Text:
            styles: { padding: 8, fontSize: 14 }
            text: "You selected toggle button 1 value: "
            id: onChangeDemo

ToggleButton2:
  body:
    Column:
      styles: { gap: 10 }
      crossAxis: center
      children:
        - Text:
            styles: { fontSize: 18 }
            text: "Toggle Button 2 - With Styles"
        - Text:
            text: "Styles like color, selectedColor, backgroundColor, selectedBackgroundColor, borderColor, selectedBorderColor, spacing, runSpacing"
        - ToggleButton:
            styles:
              {
                spacing: 10,
                runSpacing: 10,
                color: blue,
                selectedColor: white,
                backgroundColor: yellow,
                selectedBackgroundColor: green,
                selectedBorderColor: blue,
                borderColor: red,
              }
            value: Blue
            items:
              - Red
              - Blue
              - Green
              - Orange
              - Purple
              - Yellow
            onChange: |
              //@code
              onChangeDemo.text = 'You selected toggle button 2 value: ' +this.value;

        - Text:
            styles: { padding: 8, fontSize: 14 }
            text: "You selected toggle button 2 value: "
            id: onChangeDemo

ToggleButton3:
  body:
    Column:
      styles: { gap: 10 }
      crossAxis: center
      children:
        - Text:
            styles: { fontSize: 18 }
            text: "Toggle Button 3 - List of Item Objects"
        - ToggleButton:
            styles:
              {
                color: grey,
                selectedColor: blue,
                selectedBorderColor: blue,
                borderColor: grey,
              }
            value: green
            items:
              - {
                  "label": "Red",
                  "value": "red",
                  "icon": { "name": "home", "size": 24 },
                }
              - { "label": "Blue", "value": "blue", "icon": { "name": "star" } }
              - { "label": "Green", "value": "green" }
            onChange: |
              //@code
              onChangeDemo.text = 'You selected toggle button 3 value: ' +this.value;

        - Text:
            styles: { padding: 8, fontSize: 14 }
            text: "You selected toggle button 3 value: "
            id: onChangeDemo

ToggleButton4:
  body:
    Column:
      styles: { gap: 10 }
      crossAxis: center
      children:
        - Text:
            styles: { fontSize: 18 }
            text: "Toggle Button 4 - Shadow"
        - ToggleButton:
            styles:
              {
                spacing: 10,
                runSpacing: 10,
                shadowColor: grey,
                backgroundColor: white,
                selectedBackgroundColor: white,
              }
            value: 25 ml
            items:
              - 10 ml
              - 25 ml
              - 50 ml
              - 75 ml
            onChange: |
              //@code
              onChangeDemo.text = 'You selected toggle button 4 value: ' +this.value;

        - Text:
            styles: { padding: 8, fontSize: 14 }
            text: "You selected toggle button 4 value: "
            id: onChangeDemo

  ```
</details>

File: assets/schema/ensemble_schema.json (hunk 6 above)
Comment (vusters) [reply assets/schema/ensemble_schema.json:15]: should be ok. They are just json formatted differently.

**Pull request merged**